{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[],"authorship_tag":"ABX9TyOIS3/EiMXm0DyT9YOP5Fmv"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"code","execution_count":1,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"fFC08sEXf8Pz","executionInfo":{"status":"ok","timestamp":1714589660954,"user_tz":-180,"elapsed":105377,"user":{"displayName":"Ð“ÐµÐ¾Ñ€Ð³Ð¸Ð¹ Ð‘Ð¾Ð¹ÐºÐ¾Ð²","userId":"05592188241194498478"}},"outputId":"0d7ba519-178c-4e57-ab6c-6e8bfa60a5de"},"outputs":[{"output_type":"stream","name":"stdout","text":["Collecting biopython\n","  Downloading biopython-1.83-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.1 MB)\n","\u001b[2K     \u001b[90mâ”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”\u001b[0m \u001b[32m3.1/3.1 MB\u001b[0m \u001b[31m13.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hRequirement already satisfied: numpy in /usr/local/lib/python3.10/dist-packages (from biopython) (1.25.2)\n","Installing collected packages: biopython\n","Successfully installed biopython-1.83\n","â¬ Downloading https://github.com/conda-forge/miniforge/releases/download/23.11.0-0/Mambaforge-23.11.0-0-Linux-x86_64.sh...\n","ðŸ“¦ Installing...\n","ðŸ“Œ Adjusting configuration...\n","ðŸ©¹ Patching environment...\n","â² Done in 0:00:13\n","ðŸ” Restarting kernel...\n","Channels:\n"," - bioconda\n"," - conda-forge\n","Platform: linux-64\n","Collecting package metadata (repodata.json): - \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\bdone\n","Solving environment: - \b\b\\ \b\b| \b\bdone\n","\n","\n","==> WARNING: A newer version of conda exists. <==\n","    current version: 23.11.0\n","    latest version: 24.3.0\n","\n","Please update conda by running\n","\n","    $ conda update -n base -c conda-forge conda\n","\n","\n","\n","## Package Plan ##\n","\n","  environment location: /usr/local\n","\n","  added / updated specs:\n","    - seqkit\n","\n","\n","The following packages will be downloaded:\n","\n","    package                    |            build\n","    ---------------------------|-----------------\n","    ca-certificates-2024.2.2   |       hbcca054_0         152 KB  conda-forge\n","    certifi-2024.2.2           |     pyhd8ed1ab_0         157 KB  conda-forge\n","    openssl-3.3.0              |       hd590300_0         2.8 MB  conda-forge\n","    seqkit-2.8.1               |       h9ee0642_0         7.1 MB  bioconda\n","    ------------------------------------------------------------\n","                                           Total:        10.2 MB\n","\n","The following NEW packages will be INSTALLED:\n","\n","  seqkit             bioconda/linux-64::seqkit-2.8.1-h9ee0642_0 \n","\n","The following packages will be UPDATED:\n","\n","  ca-certificates                     2023.11.17-hbcca054_0 --> 2024.2.2-hbcca054_0 \n","  certifi                           2023.11.17-pyhd8ed1ab_0 --> 2024.2.2-pyhd8ed1ab_0 \n","  openssl                                  3.2.0-hd590300_1 --> 3.3.0-hd590300_0 \n","\n","\n","\n","Downloading and Extracting Packages:\n","seqkit-2.8.1         | 7.1 MB    | :   0% 0/1 [00:00<?, ?it/s]\n","openssl-3.3.0        | 2.8 MB    | :   0% 0/1 [00:00<?, ?it/s]\u001b[A\n","\n","certifi-2024.2.2     | 157 KB    | :   0% 0/1 [00:00<?, ?it/s]\u001b[A\u001b[A\n","\n","\n","ca-certificates-2024 | 152 KB    | :   0% 0/1 [00:00<?, ?it/s]\u001b[A\u001b[A\u001b[A\n","\n","\n","seqkit-2.8.1         | 7.1 MB    | :   0% 0.0022029514763095993/1 [00:00<00:57, 57.53s/it]\n","\n","certifi-2024.2.2     | 157 KB    | :  10% 0.10204348557228184/1 [00:00<00:01,  1.22s/it]\u001b[A\u001b[A\n","\n","\n","ca-certificates-2024 | 152 KB    | : 100% 1.0/1 [00:00<00:00,  1.02it/s]                \u001b[A\u001b[A\u001b[A\n","openssl-3.3.0        | 2.8 MB    | :   1% 0.005659047239435657/1 [00:00<00:28, 29.08s/it]\u001b[A\n","\n","certifi-2024.2.2     | 157 KB    | : 100% 1.0/1 [00:00<00:00,  1.22s/it]                \u001b[A\u001b[A\n","openssl-3.3.0        | 2.8 MB    | : 100% 1.0/1 [00:00<00:00,  2.79it/s]                 \u001b[A\n","                                                                        \n","                                                                        \u001b[A\n","\n","                                                                        \u001b[A\u001b[A\n","\n","\n","                                                                        \u001b[A\u001b[A\u001b[A\n","Preparing transaction: - \b\bdone\n","Verifying transaction: | \b\bdone\n","Executing transaction: - \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\bdone\n"]}],"source":["! pip install biopython\n","! pip install -q condacolab\n","import condacolab\n","condacolab.install()\n","! conda install -c bioconda seqkit"]},{"cell_type":"code","source":["from pydrive2.auth import GoogleAuth\n","from pydrive2.drive import GoogleDrive\n","from google.colab import auth\n","from oauth2client.client import GoogleCredentials\n","# Authenticate and create the PyDrive client.\n","# This only needs to be done once per notebook.\n","auth.authenticate_user()\n","gauth = GoogleAuth()\n","gauth.credentials = GoogleCredentials.get_application_default()\n","drive = GoogleDrive(gauth)"],"metadata":{"id":"9xo--PdIgpsb","executionInfo":{"status":"ok","timestamp":1714589724388,"user_tz":-180,"elapsed":11850,"user":{"displayName":"Ð“ÐµÐ¾Ñ€Ð³Ð¸Ð¹ Ð‘Ð¾Ð¹ÐºÐ¾Ð²","userId":"05592188241194498478"}}},"execution_count":1,"outputs":[]},{"cell_type":"code","source":["# https://drive.google.com/file/d/1XZAwTCcIlANrd1wwR12_rbi1NvaQINos/view?usp=drive_link\n","# https://drive.google.com/file/d/1_dl0q_4JJe8IJxB1L2qDGc445gVlhWN-/view?usp=drive_link\n","# https://drive.google.com/file/d/14Ckpl7W6zfItYvDUmvm7lqbkd48CG20i/view?usp=drive_link\n","id_links = {\"hw_file1.fasta\":\"1XZAwTCcIlANrd1wwR12_rbi1NvaQINos\",\n","            \"hw_file2.fasta\":\"1_dl0q_4JJe8IJxB1L2qDGc445gVlhWN-\",\n","            \"hw_file3.fasta\":\"14Ckpl7W6zfItYvDUmvm7lqbkd48CG20i\"}\n","for name, file_id in id_links.items():\n","  downloaded = drive.CreateFile({'id': file_id})\n","  downloaded.GetContentFile(name)"],"metadata":{"id":"HIiGjst2g2hZ","executionInfo":{"status":"ok","timestamp":1714590060365,"user_tz":-180,"elapsed":60604,"user":{"displayName":"Ð“ÐµÐ¾Ñ€Ð³Ð¸Ð¹ Ð‘Ð¾Ð¹ÐºÐ¾Ð²","userId":"05592188241194498478"}}},"execution_count":5,"outputs":[]},{"cell_type":"code","source":["import requests\n","import re\n","import json\n","from Bio import SeqIO\n","import subprocess\n","import sys\n","import argparse"],"metadata":{"id":"xaHYWYQogEs7","executionInfo":{"status":"ok","timestamp":1714590100655,"user_tz":-180,"elapsed":416,"user":{"displayName":"Ð“ÐµÐ¾Ñ€Ð³Ð¸Ð¹ Ð‘Ð¾Ð¹ÐºÐ¾Ð²","userId":"05592188241194498478"}}},"execution_count":6,"outputs":[]},{"cell_type":"code","source":["class BioPythonLib:\n","    def __init__(self, file_name):\n","        self.regex = {\"Protein\": \"^.*([OPQ][0-9][A-Z0-9]{3}[0-9]|[A-NR-Z][0-9]([A-Z][A-Z0-9]{2}[0-9]){1,2}).*$\",\n","                      \"DNA\": \"^.*(ENS[A-Z]{0,4}|MGP_[A-Za-z0-9]{0,10}_)(E|FM|G|GT|P|R|T)(\\d{11}).*$\"}\n","        self.filename = file_name\n","        self.SeqIO_ids = []\n","        self.output = {\"seqkit_result\": self.seqkit_stats()}\n","        self.biopython_parser()\n","        self.access_database()\n","        self.show_output(self.output)\n","\n","    def seqkit_stats(self):\n","        seqkit = subprocess.run((\"seqkit\", \"stats\", self.filename, \"-a\"),\n","                                capture_output=True,\n","                                text=True)\n","\n","        # If error arises\n","        if (seqkit.stderr != ''):\n","            print(\"Error:\", seqkit.stderr)\n","            sys.exit(1)\n","\n","        # Output\n","        if (seqkit.stdout == ''):\n","            print(\"Output is empty. Check the fasta file\")\n","            sys.exit(1)\n","\n","        seqkit_out = seqkit.stdout.strip().split('\\n')\n","\n","        # split names and values\n","        prop_names = seqkit_out[0].split()[1:]\n","        prop_vals = seqkit_out[1].split()[1:]\n","\n","        seq_result = dict(zip(prop_names, prop_vals))\n","        # using zip\n","        self.seqkit_result = {\"fasta_seqkit_stat_info\": seq_result,\n","                              \"fasta_type\": seq_result['type'],\n","                              \"fasta_num_seqs\": int(seq_result['num_seqs']),\n","                              }\n","\n","        return self.seqkit_result\n","\n","    def biopython_parser(self):\n","        sequences = SeqIO.parse(self.filename, 'fasta')  # seq input; returns an iterator\n","        regex = self.regex[self.seqkit_result[\"fasta_type\"]]\n","\n","        for seq in sequences:\n","\n","            seq_description = seq.description\n","            seq_sequence = seq.seq\n","\n","            # seq_id\n","            match = re.fullmatch(regex, seq_description)\n","\n","            if match:\n","                if self.seqkit_result[\"fasta_type\"] == 'DNA':\n","                    id_chunks = re.findall(regex, seq_description)\n","                    seq_id = [''.join(chunk) for chunk in id_chunks][0]\n","\n","                if self.seqkit_result[\"fasta_type\"] == 'Protein':\n","                    seq_id = re.findall(regex, seq_description)[0][0]\n","\n","                self.SeqIO_ids.append(seq_id)\n","                self.output[f'seq_id_{seq_id}_info'] = {\"description\": seq_description, \"sequence\": seq_sequence}\n","\n","            else:\n","                print(\"No ID match found.\")\n","                sys.exit(1)\n","\n","    def get_uniprot(self, ids: list):\n","        accessions = ','.join(ids)\n","        endpoint = \"https://rest.uniprot.org/uniprotkb/accessions\"\n","        http_function = requests.get\n","        http_args = {'params': {'accessions': accessions}}\n","\n","        return http_function(endpoint, **http_args)\n","\n","    def get_ensembl(self, ids: list):\n","        server = \"https://rest.ensembl.org\"\n","        ext = \"/lookup/id\"\n","        headers = {\"Content-Type\": \"application/json\"}\n","        x = {\"ids\": ids}\n","        y = json.dumps(x)\n","        resp = requests.post(server + ext, headers=headers, data=y)\n","        return resp\n","\n","    def parse_response_uniprot(self, resp: dict):\n","        resp = resp.json()\n","        resp = resp[\"results\"]\n","        output = {}\n","        for val in resp:\n","            acc = val['primaryAccession']\n","            species = val['organism']['scientificName']\n","            gene = val['genes']\n","            seq = val['sequence']\n","            output[acc] = {'organism': species, 'geneInfo': gene, 'sequenceInfo': seq, 'type': 'protein'}\n","\n","        return output\n","\n","    def parse_response_ensembl(self, resp: dict):\n","        resp = resp.json()\n","        output = {}\n","        for val in resp.values():\n","            id = val['id']\n","            species = val['species']\n","            gene = {\n","                'assembly_name': val['assembly_name'],\n","                'biotype': val['biotype'],\n","                'canonical_transcript': val['canonical_transcript']\n","            }\n","            seq = {\n","                'seq_region_name': val['seq_region_name'],\n","                'start': val['start'],\n","                'end': val['end']\n","            }\n","            object_type = val['object_type']\n","            output[id] = {'organism': species,\n","                          'geneInfo': gene,\n","                          'sequenceInfo': seq,\n","                          'type': object_type}\n","\n","        return output\n","\n","    def access_database(self):\n","        ids = self.SeqIO_ids\n","        if self.seqkit_result[\"fasta_type\"] == 'Protein':\n","            data = self.parse_response_uniprot(self.get_uniprot(ids))\n","            self.output[\"DB_name\"] = \"uniprot\"\n","            self.output[\"DB_result\"] = data\n","        elif self.seqkit_result[\"fasta_type\"] == 'DNA':\n","            data = self.parse_response_ensembl(self.get_ensembl(ids))\n","            self.output[\"DB_name\"] = \"ENSEMBL\"\n","            self.output[\"DB_result\"] = data\n","\n","    def show_output(self, v, indent=0):\n","        for key, value in v.items():\n","            print('\\t' * indent + str(key))\n","            if isinstance(value, dict):\n","                self.show_output(value, indent + 1)\n","            else:\n","                print('\\t' * (indent + 1) + str(value))"],"metadata":{"id":"wsyMRyjjohj1","executionInfo":{"status":"ok","timestamp":1714595929268,"user_tz":-180,"elapsed":229,"user":{"displayName":"Ð“ÐµÐ¾Ñ€Ð³Ð¸Ð¹ Ð‘Ð¾Ð¹ÐºÐ¾Ð²","userId":"05592188241194498478"}}},"execution_count":29,"outputs":[]},{"cell_type":"code","source":["file_name = \"hw_file2.fasta\"\n","bio = BioPythonLib(file_name)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"DqKZ92vrieRy","executionInfo":{"status":"ok","timestamp":1714595934454,"user_tz":-180,"elapsed":992,"user":{"displayName":"Ð“ÐµÐ¾Ñ€Ð³Ð¸Ð¹ Ð‘Ð¾Ð¹ÐºÐ¾Ð²","userId":"05592188241194498478"}},"outputId":"2a16828e-4308-4ef5-9418-4e52b65c0bb1"},"execution_count":30,"outputs":[{"output_type":"stream","name":"stdout","text":["seqkit_result\n","\tfasta_seqkit_stat_info\n","\t\tformat\n","\t\t\tFASTA\n","\t\ttype\n","\t\t\tDNA\n","\t\tnum_seqs\n","\t\t\t6\n","\t\tsum_len\n","\t\t\t86\n","\t\tmin_len\n","\t\t\t9\n","\t\tavg_len\n","\t\t\t14.3\n","\t\tmax_len\n","\t\t\t23\n","\t\tQ1\n","\t\t\t10\n","\t\tQ2\n","\t\t\t13.5\n","\t\tQ3\n","\t\t\t17\n","\t\tsum_gap\n","\t\t\t0\n","\t\tN50\n","\t\t\t16\n","\t\tN50_num\n","\t\t\t3\n","\t\tQ20(%)\n","\t\t\t0\n","\t\tQ30(%)\n","\t\t\t0\n","\t\tAvgQual\n","\t\t\t0\n","\t\tGC(%)\n","\t\t\t45.35\n","\tfasta_type\n","\t\tDNA\n","\tfasta_num_seqs\n","\t\t6\n","seq_id_ENSMUSG00000096749_info\n","\tdescription\n","\t\tENSMUST00000196221.2 cds chromosome:GRCm39:14:54350925:54350933:1 gene:ENSMUSG00000096749.3 gene_biotype:TR_D_gene transcript_biotype:TR_D_gene gene_symbol:Trdd1 description:T cell receptor delta diversity 1 [Source:MGI Symbol;Acc:MGI:4439547]\n","\tsequence\n","\t\tATGGCATAT\n","seq_id_ENSMUSG00000096176_info\n","\tdescription\n","\t\tENSMUST00000177564.2 cds chromosome:GRCm39:14:54359683:54359698:1 gene:ENSMUSG00000096176.2 gene_biotype:TR_D_gene transcript_biotype:TR_D_gene gene_symbol:Trdd2 description:T cell receptor delta diversity 2 [Source:MGI Symbol;Acc:MGI:4439546]\n","\tsequence\n","\t\tATCGGAGGGATACGAG\n","seq_id_ENSG00000271336_info\n","\tdescription\n","\t\tENST00000605284.1 cds chromosome:GRCh38:15:20011153:20011169:-1 gene:ENSG00000271336.1 gene_biotype:IG_D_gene transcript_biotype:IG_D_gene gene_symbol:IGHD1OR15-1A description:immunoglobulin heavy diversity 1/OR15-1A (non-functional) [Source:HGNC Symbol;Acc:HGNC:5487]\n","\tsequence\n","\t\tGGTATAACTGGAACAAC\n","seq_id_ENSG00000270961_info\n","\tdescription\n","\t\tENST00000604642.1 cds chromosome:GRCh38:15:20003840:20003862:-1 gene:ENSG00000270961.1 gene_biotype:IG_D_gene transcript_biotype:IG_D_gene gene_symbol:IGHD5OR15-5A description:immunoglobulin heavy diversity 5/OR15-5A (non-functional) [Source:HGNC Symbol;Acc:HGNC:5512]\n","\tsequence\n","\t\tGTGGATATAGTGTCTACGATTAC\n","seq_id_ENSDARG00000116509_info\n","\tdescription\n","\t\tENSDART00000189431.1 cds chromosome:GRCz11:2:36087769:36087779:1 gene:ENSDARG00000116509.1 gene_biotype:TR_D_gene transcript_biotype:TR_D_gene gene_symbol:BX681417.25\n","\tsequence\n","\t\tGATTGGGGTAC\n","seq_id_ENSDARG00000116470_info\n","\tdescription\n","\t\tENSDART00000189226.1 cds chromosome:GRCz11:2:36088047:36088056:1 gene:ENSDARG00000116470.1 gene_biotype:TR_D_gene transcript_biotype:TR_D_gene gene_symbol:BX681417.24\n","\tsequence\n","\t\tTCTGGACTAC\n","DB_name\n","\tENSEMBL\n","DB_result\n","\tENSG00000271336\n","\t\tspeciesInfo\n","\t\t\thomo_sapiens\n","\t\tgeneInfo\n","\t\t\tbiotype\n","\t\t\t\tIG_D_gene\n","\t\ttranscriptInfo\n","\t\t\tassembly_name\n","\t\t\t\tGRCh38\n","\t\t\tcanonical_transcript\n","\t\t\t\tENST00000605284.1\n","\t\t\tstart\n","\t\t\t\t20011153\n","\t\t\tend\n","\t\t\t\t20011169\n","\t\tsequenceInfo\n","\t\t\tassembly_name\n","\t\t\t\tGRCh38\n","\t\t\tstart\n","\t\t\t\t20011153\n","\t\t\tend\n","\t\t\t\t20011169\n","\tENSDARG00000116470\n","\t\tspeciesInfo\n","\t\t\tdanio_rerio\n","\t\tgeneInfo\n","\t\t\tbiotype\n","\t\t\t\tTR_D_gene\n","\t\ttranscriptInfo\n","\t\t\tassembly_name\n","\t\t\t\tGRCz11\n","\t\t\tcanonical_transcript\n","\t\t\t\tENSDART00000189226.1\n","\t\t\tstart\n","\t\t\t\t36088047\n","\t\t\tend\n","\t\t\t\t36088056\n","\t\tsequenceInfo\n","\t\t\tassembly_name\n","\t\t\t\tGRCz11\n","\t\t\tstart\n","\t\t\t\t36088047\n","\t\t\tend\n","\t\t\t\t36088056\n","\tENSDARG00000116509\n","\t\tspeciesInfo\n","\t\t\tdanio_rerio\n","\t\tgeneInfo\n","\t\t\tbiotype\n","\t\t\t\tTR_D_gene\n","\t\ttranscriptInfo\n","\t\t\tassembly_name\n","\t\t\t\tGRCz11\n","\t\t\tcanonical_transcript\n","\t\t\t\tENSDART00000189431.1\n","\t\t\tstart\n","\t\t\t\t36087769\n","\t\t\tend\n","\t\t\t\t36087779\n","\t\tsequenceInfo\n","\t\t\tassembly_name\n","\t\t\t\tGRCz11\n","\t\t\tstart\n","\t\t\t\t36087769\n","\t\t\tend\n","\t\t\t\t36087779\n","\tENSMUSG00000096749\n","\t\tspeciesInfo\n","\t\t\tmus_musculus\n","\t\tgeneInfo\n","\t\t\tbiotype\n","\t\t\t\tTR_D_gene\n","\t\ttranscriptInfo\n","\t\t\tassembly_name\n","\t\t\t\tGRCm39\n","\t\t\tcanonical_transcript\n","\t\t\t\tENSMUST00000196221.2\n","\t\t\tstart\n","\t\t\t\t54350925\n","\t\t\tend\n","\t\t\t\t54350935\n","\t\tsequenceInfo\n","\t\t\tassembly_name\n","\t\t\t\tGRCm39\n","\t\t\tstart\n","\t\t\t\t54350925\n","\t\t\tend\n","\t\t\t\t54350935\n","\tENSG00000270961\n","\t\tspeciesInfo\n","\t\t\thomo_sapiens\n","\t\tgeneInfo\n","\t\t\tbiotype\n","\t\t\t\tIG_D_gene\n","\t\ttranscriptInfo\n","\t\t\tassembly_name\n","\t\t\t\tGRCh38\n","\t\t\tcanonical_transcript\n","\t\t\t\tENST00000604642.1\n","\t\t\tstart\n","\t\t\t\t20003840\n","\t\t\tend\n","\t\t\t\t20003862\n","\t\tsequenceInfo\n","\t\t\tassembly_name\n","\t\t\t\tGRCh38\n","\t\t\tstart\n","\t\t\t\t20003840\n","\t\t\tend\n","\t\t\t\t20003862\n","\tENSMUSG00000096176\n","\t\tspeciesInfo\n","\t\t\tmus_musculus\n","\t\tgeneInfo\n","\t\t\tbiotype\n","\t\t\t\tTR_D_gene\n","\t\ttranscriptInfo\n","\t\t\tassembly_name\n","\t\t\t\tGRCm39\n","\t\t\tcanonical_transcript\n","\t\t\t\tENSMUST00000177564.2\n","\t\t\tstart\n","\t\t\t\t54359683\n","\t\t\tend\n","\t\t\t\t54359698\n","\t\tsequenceInfo\n","\t\t\tassembly_name\n","\t\t\t\tGRCm39\n","\t\t\tstart\n","\t\t\t\t54359683\n","\t\t\tend\n","\t\t\t\t54359698\n"]}]}]}